Math 4e Cahier des compétances

page 29, parcours rouge

**32. Pour les experts** Est-tu capable de calculer la somme des chiffre du nombre : 

In [1]:
%reset -f
n = 10**2018 - 2018
print('n = ', n)

n =  99999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999999

Compter la somme des chifres du nombre.

In [2]:
total = 0                     # set total to zero
while(n > 0):
    last_dig = n % 10         # take last digit
    total = total + last_dig  # add last digit to total
    n = n // 10               # delete the last digit
print("The total sum of digits is:", total)

The total sum of digits is: 18152


User must first enter the value *n*.

- Line 1. The `total` is set to zero.
- Line 2. Loop `while n > 0` 
- Line 3. The last digit of the number is obtained by using the modulus operator.
- Line 4. The digit is added to *total* variable each time the loop is executed.
- Line 5. The number is devided by 10 and taken only the integer part (*floor division*).
- Line 6. When the loop is finished, the total sum of the number is printed.

Tester *modulo* et *division entière* :

In [3]:
9 % 10

9

In [4]:
9 // 10

0

Calcul

On fait des soustractions :

In [5]:
1e4 - 2018

7982.0

In [6]:
1e5 - 2018

97982.0

In [7]:
1e6 - 2018

997982.0

Pour $10^n - 2018, n > 4$ on constate que :
- le résultat a $n$ chiffres ;
- les denières chiffres sont 7982 ;
- les reste de $n - 4$ chifres sont 9.



Pour $n = 2018$ on obtient $10^{2018} - 2018$. La somme des chiffres de ce nombre est :

In [8]:
n = 2018
somme_chiffres = (n - 4) * 9 + 7 + 9 + 8 + 2
print(somme_chiffres)

18152
